File: 20. Parenthesis Matching.cpp
#include <iostream>
#include <string>
#include <stack>

bool isMatched(std::string s)
{
    if (s.size() % 2 != 0)
        return false;
    std::stack<char> tmp;
    for (char &c : s)
    {
        switch (c)
        {
        case '(':
        case '[':
        case '{':
            tmp.push(c);
            break;
        case ')':
            if (tmp.empty() || tmp.top() != '(')
                return false;
            else
                tmp.pop();
            break;
        case ']':
            if (tmp.empty() || tmp.top() != '[')
                return false;
            else
                tmp.pop();
            break;
        case '}':
            if (tmp.empty() || tmp.top() != '{')
                return false;
            else
                tmp.pop();
            break;
        default:;
        }
    }
    return tmp.empty();
}

int main()
{
    std::string s1 = "(){[[()]]}{}";
    std::string s2 = "[]((())}";
    std::cout << isMatched(s1) << std::endl;
    std::cout << isMatched(s2) << std::endl;
}

File: pyversion/20. Parenthesis Matching.py
def isMatched(s):
    if len(s) % 2 != 0:
        return False
    stack = []
    par_dict = {'}': '{', ']': '[', ')': '('}
    for char in s:
        if char in par_dict.values():
            stack.append(char)
        elif char in par_dict.keys():
            if stack == [] or par_dict[char] != stack.pop():
                return False
        else:
            return False
    return stack == []


if __name__ == '__main__':
    print(isMatched('()()(){{{{[]}}}}'))
    print(isMatched('((())}'))
